yarn-project/prover-client/src/prover/bb_prover.ts

```diff
@@ -233,27 +282,110 @@ export class BBNativeRollupProver implements CircuitProver {
+    const outputWitnessFile = `${bbWorkingDirectory}/partial-witness.gz`;
+
+    // Generate the partial witness using the ACVM
+    // A further temp directory will be created beneath ours and then cleaned up after the partial witness has been copied to our specified location
+    const simulator = new NativeACVMSimulator(
+      this.config.acvmWorkingDirectory,
+      this.config.acvmBinaryPath,
+      outputWitnessFile,
+    );
+
+    const artifact = ServerCircuitArtifacts[circuitType];
+
+    logger.debug(`Generating witness data for ${circuitType}`);
+
+    const outputWitness = await simulator.simulateCircuit(witnessMap, artifact);
+
+    const outputType = convertOutput(outputWitness);
+
+    // Now prove the circuit from the generated witness
+    logger.debug(`Proving ${circuitType}...`);
+
+    const provingResult = await generateProof(
+      this.config.bbBinaryPath,
+      bbWorkingDirectory,
+      circuitType,
+      artifact,
+      outputWitnessFile,
+      logger.debug,
+    );
+
+    if (provingResult.status === BB_RESULT.FAILURE) {
+      logger.error(`Failed to generate proof for ${circuitType}: ${provingResult.reason}`);
+      throw new Error(provingResult.reason);
+    }
+
+    // Ensure our vk cache is up to date
+    await this.updateVerificationKeyAfterProof(provingResult.vkPath!, circuitType);
+
+    // Read the proof and then cleanup up our temporary directory
+    const proof = await this.readProofAsFields<PROOF_LENGTH>(provingResult.proofPath!, circuitType);
 
     await fs.rm(bbWorkingDirectory, { recursive: true, force: true });
 
     logger.info(
-      `Generated proof for ${circuitType} in ${provingResult.duration} ms, size: ${proofBuffer.length} bytes`,
+      `Generated proof for ${circuitType} in ${provingResult.duration} ms, size: ${proof.proof.length} fields`,
     );
 
-    return [outputWitness, new Proof(proofBuffer)];
+    return [outputType, proof];
   }
 
+  /**
+   * Verifies a proof, will generate the verification key if one is not cached internally
+   * @param circuitType - The type of circuit whose proof is to be verified
+   * @param proof - The proof to be verified
+   */
   public async verifyProof(circuitType: ServerProtocolArtifact, proof: Proof) {
     // Create random directory to be used for temp files
     const bbWorkingDirectory = `${this.config.bbWorkingDirectory}/${randomBytes(8).toString('hex')}`;
     await fs.mkdir(bbWorkingDirectory, { recursive: true });
 
     const proofFileName = `${bbWorkingDirectory}/proof`;
-    const verificationKeyPath = this.verificationKeyDirectories.get(circuitType);
+    const verificationKeyPath = `${bbWorkingDirectory}/vk`;
+    const verificationKey = await this.getVerificationKeyDataForCircuit(circuitType);
+
+    logger.debug(`Verifying with key: ${verificationKey.hash.toString()}`);
+
+    if (!verificationKey) {
+      throw new Error(`Unable to verify proof for circuit ${circuitType}, verification key not available`);
+    }
```

Comment:
Is this a left over? I don't think this can ever trigger because `getVerificationKeyDataForCircuit` throws if it can't generate the key.

Reply:
Yes, removed